### PR TITLE
Replace boost filesystem with std filesystem

### PR DESCRIPTION
**Pull request by WhiredPlanck**
https://github.com/rime/librime/pull/649 is to do so.

### CODE REVIEW

**Comment by WhiredPlanck** [comment]
@hchunhui Since librime doesn’t merge the PR yet, the CI tests here will not be succeed. And you can see that https://github.com/rime/librime/pull/649 also failed the CI test of docker due to a build error from librime-lua, I have no idea how to fix it ......

**Comment by hchunhui** [comment]
> @hchunhui Since librime doesn’t merge the PR yet, the CI tests here will not be succeed. And you can see that [rime/librime#649](https://github.com/rime/librime/pull/649) also failed the CI test of docker due to a build error from librime-lua, I have no idea how to fix it ......

I have solved the build error in master branch, please rerun CI test in rime/librime#649 .

**Comment by WhiredPlanck** [comment]
@hchunhui Cool! So much thanks!

**Comment by WhiredPlanck** [comment]
@hchunhui https://github.com/rime/librime/pull/649 still failed the Docker CI test due to boost filesystem not found. I think it needs the changes from here.

**Comment by hchunhui** [comment]
@WhiredPlanck The master branch should fix the CI error.

And to ensure a smooth transition, please add proper `#if` guard, e.g.
```c++
#if __cplusplus >= 201703L || _MSVC_LANG >= 201703L
// std....
#else
// boost...
#endif
```

**Comment by WhiredPlanck** [comment]
@hchunhui Will fix it soon.

File: src/lib/luatype_boost_optional.h
Comment (hchunhui) [inline src/lib/luatype_boost_optional.h:1]: The change is incorrect and unnecessary. the `optional` ambiguity has been fixed by master branch.

Comment (WhiredPlanck) [reply src/lib/luatype_boost_optional.h:1]: Wow, you are really fast. Thanks, I will revert it soon.

**Comment by WhiredPlanck** [comment]
@hchunhui Done.

**Comment by hchunhui** [comment]
Merged. Thanks!

**Pull request merged**